File: lib/data/repository/contract/comaps_navigation_repository.dart
import '../../../core/result/result.dart';
import '../../../domain/model/comaps_navigation.dart';

/// Live navigation context borrowed from CoMaps while OpenVitals records.
///
/// The ownership boundary is the whole point of this feature: **CoMaps plans and
/// navigates; OpenVitals records.** Nothing here can start, stop or steer a
/// route — it reads what CoMaps is already doing, and nothing it returns is ever
/// written to Health Connect.
abstract interface class CoMapsNavigationRepository {
  /// What CoMaps can tell us right now. Every outcome is a [CoMapsNavigationState],
  /// including the unavailable ones — an `Err` here means the *bridge* failed,
  /// not that guidance is unavailable, which is an ordinary answer.
  Future<Result<CoMapsNavigationState>> readLive();

  Future<Result<bool>> hasPermission();

  /// Asks for `app.comaps.permission.READ_NAVIGATION_DATA`. False when the user
  /// refuses, and also when CoMaps is not installed — the permission is
  /// CoMaps', and does not exist without it.
  Future<Result<bool>> requestPermission();

  Future<Result<bool>> canLaunchCoMaps();

  /// Opens CoMaps so the user can plan a route, centred on their last fix when
  /// there is one.
  Future<Result<bool>> launchForPlanning({double? latitude, double? longitude});

  /// The guidance sampled during a recording, kept against the saved activity.
  /// App-local history only.
  Future<Result<void>> saveSamples(
    String activityId,
    List<CoMapsNavigationSnapshot> samples,
  );

  Future<Result<List<CoMapsNavigationSnapshot>>> loadSamples(String activityId);
}

File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/CoMapsNavigationRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import kotlinx.coroutines.flow.Flow
import tech.mmarca.openvitals.domain.model.CoMapsNavigationSnapshot
import tech.mmarca.openvitals.domain.model.CoMapsNavigationState
import tech.mmarca.openvitals.domain.model.CoMapsRoutePolyline

/**
 * What OpenVitals can learn from a navigating CoMaps, and the samples kept
 * per activity. Read-only; nothing is written to Health Connect.
 */
interface CoMapsNavigationRepository {
    /** One reading of the live row, never throwing — failures become [CoMapsNavigationState.Error]. */
    suspend fun readLive(): CoMapsNavigationState

    /** The live feed: an initial reading, then one per `notifyChange`. */
    fun watchLive(): Flow<CoMapsNavigationState>

    /** The followed route's polyline, or null when there is none to read. */
    suspend fun readRouteGeometry(revision: Int): CoMapsRoutePolyline?

    /** The flavour-specific runtime permission to request, null without a CoMaps installed. */
    fun permissionName(): String?

    fun hasPermission(): Boolean

    /** Call after a grant result: the package probe may now answer differently. */
    fun onPermissionChanged()

    fun canLaunchCoMaps(): Boolean

    fun launchForPlanning(latitude: Double?, longitude: Double?): Boolean

    fun saveSamples(activityId: String, samples: List<CoMapsNavigationSnapshot>)

    fun loadSamples(activityId: String): List<CoMapsNavigationSnapshot>

    fun deleteSamples(activityId: String)
}
